File: regression_songs.py
# pandas : Python Data Analysis Library
import pandas as pd
from pandas import Series,DataFrame
import numpy as np

# Read CSV file into DataFrame
songs_df = pd.read_csv("songs.csv")

#take a quick look into "songs.csv" 
print(songs_df.head())
print(songs_df.tail(5))

# Songs of 2010
ans_1_1=songs_df[songs_df.year==2010].count()["year"] 
print(ans_1_1)
# Songs from "Michael Jackson"
ans_1_2=songs_df[songs_df.artistname=="Michael Jackson"].count()["artistname"]
print(ans_1_2)
# Michael Jackson & Top10
ans_1_3=songs_df.loc[(songs_df.artistname=="Michael Jackson")&(songs_df.Top10==1)]["songtitle"]
print(ans_1_3)
# timesignature values
ans_1_4_1=Series(songs_df.timesignature.values.ravel()).unique()
print(ans_1_4_1)
# most frequent timesignature 
ans_1_4_2=songs_df.groupby(songs_df.timesignature, sort=False).count()
print(ans_1_4_2)
# song with the highest tempo
max_tempo_idx=songs_df.tempo.idxmax(axis=0)
ans_1_5=songs_df.songtitle[max_tempo_idx]
print(ans_1_5)


#-----------------------#
#create prediction model#
#-----------------------#

# prepare train data & test data
nonvars=["year", "songtitle", "artistname", "songID", "artistID"]
'''
# Way 1: 
SongTrain=songs_df.drop(songs_df[songs_df.year==2010].index)
SongTrain=SongTrain.drop(nonvars, 1)
SongTest=songs_df.drop(songs_df[songs_df.year!=2010].index)
SongTest=SongTest.drop(nonvars, 1)
'''

# training data (1990~2009) 
# Top10:nonTop10 = 1000:1000
songs_df_train=songs_df.drop(songs_df[songs_df.year==2010].index).drop(nonvars, 1)
songs_df_Top10=songs_df_train.drop(songs_df_train[songs_df_train.Top10==0].index)
songs_df_not_Top10=songs_df_train.drop(songs_df_train[songs_df_train.Top10==1].index)
SongTrain=songs_df_Top10.sample(1000).append(songs_df_not_Top10.sample(1000))




# testing data (2010)
songs_df_test=songs_df.drop(songs_df[songs_df.year!=2010].index).drop(nonvars, 1)
SongTest=songs_df_test.sample(373)





# Create Model
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
logreg = LogisticRegression()
Y_ans=SongTest.Top10

# Model 1
X_train=SongTrain.drop("Top10",1)
Y_train=SongTrain.Top10
X_test1=SongTest.drop("Top10",1)
SongLog_1_score=logreg.fit(X_train, Y_train).score(X_test1, Y_ans)
print(SongLog_1_score)
Y_pred_1=logreg.predict(X_test1)
SongLog_1=logreg.fit(X_train, Y_train)
coeff_df_1 = DataFrame(X_train.columns)
coeff_df_1.columns = ['Features']
coeff_df_1["Coefficient Estimate"] = pd.Series(SongLog_1.coef_[0])
#print(coeff_df_1)
print(confusion_matrix(Y_ans, Y_pred_1))

# Model 2
SongLog_2_score=logreg.fit(X_train.drop("loudness", 1), Y_train).score(X_test1.drop("loudness",1), Y_ans)
print(SongLog_2_score)
Y_pred_2=logreg.predict(X_test1.drop("loudness",1))
SongLog_2=logreg.fit(X_train.drop("loudness", 1), Y_train)
coeff_df_2 = DataFrame(X_train.drop("loudness", 1).columns)
coeff_df_2.columns = ['Features']
coeff_df_2["Coefficient Estimate"] = pd.Series(SongLog_2.coef_[0])
#print(coeff_df_2)
print(confusion_matrix(Y_ans, Y_pred_2))

# Model 3
SongLog_3_score=logreg.fit(X_train.drop("energy", 1), Y_train).score(X_test1.drop("energy",1), Y_ans)
print(SongLog_3_score)
Y_pred_3=logreg.predict(X_test1.drop("energy",1))
SongLog_3=logreg.fit(X_train.drop("energy", 1), Y_train)
coeff_df_3 = DataFrame(X_train.drop("energy", 1).columns)
coeff_df_3.columns = ['Features']
coeff_df_3["Coefficient Estimate"] = pd.Series(SongLog_3.coef_[0])
#print(coeff_df_3)
print(confusion_matrix(Y_ans, Y_pred_3))

File: regression_songs.R
# Read CSV file into DataFrame 
songs = read.csv('songs.csv')

#take a quick look into "songs.csv"
head(songs)


# Songs of 2010
sum(songs$year==2010)
table(songs$year)
# Songs from "Michael Jackson"
sum(songs$artistname == 'Michael Jackson')
MJ = subset(songs,artistname == 'Michael Jackson')
# Michael Jackson & Top10
subset(MJ,Top10==1)
# timesignature values
unique(songs$timesignature)
# most frequent timesignature 
table(songs$timesignature)
# song with the highest tempo
songs$songtitle[which.max(songs$tempo)]






#-----------------------#
#create prediction model#
#-----------------------#

# prepare train data & test data
nonvars = c("year","songtitle","artistname","songID","artistID")
# training data (1990~2009)
SongsTrain = subset(songs,year<=2009)

library(dplyr)
# Top10:nonTop10 = 1000:1000
SongTrain1 = subset(SongsTrain,Top10==0)
SongTrain2 = subset(SongsTrain,Top10==1)
s1 = sample_n(SongTrain1,1000)
s2 = sample_n(SongTrain2,1000)
newTrain = rbind(s1,s2)


# testing data (2010)
SongsTest = subset(songs,year==2010)
# eliminate features we don't use
newTrain = newTrain[,!(names(newTrain) %in% nonvars)]
SongsTrain = SongsTrain[,!(names(SongsTrain) %in% nonvars)]
SongsTest = SongsTest[,!(names(SongsTest) %in% nonvars)]


# Create Model
# Method one: use SongsTrain
SongsLog1 = glm(Top10 ~ ., data=SongsTrain, family=binomial)
SongsLog2 = glm(Top10 ~ .-loudness,data=SongsTrain, family=binomial)
SongsLog3 = glm(Top10 ~ .-energy, data=SongsTrain, family=binomial)

# Method two: use newTrain
SongsLog1 = glm(Top10 ~ ., data=newTrain, family=binomial)
SongsLog2 = glm(Top10 ~ .-loudness,data=newTrain, family=binomial)
SongsLog3 = glm(Top10 ~ .-energy, data=newTrain, family=binomial)

# Evaluation : confusion matrix
testPredict = predict(SongsLog3, newdata=SongsTest, type="response")
confusion = table(SongsTest$Top10, testPredict >= 0.5) #row為實際，column為預測方向(右邊參數為判斷T or F)
sensitivity = confusion[4]/(confusion[2]+confusion[4])
specificity = confusion[1]/(confusion[1]+confusion[3])





SongsLog4 = glm(formula = Top10 ~ timesignature_confidence + loudness + key_confidence + 
      pitch + timbre_0_min + timbre_0_max + timbre_1_min + timbre_2_min + 
      timbre_3_max + timbre_4_min + timbre_4_max + timbre_5_min + 
      timbre_6_min + timbre_6_max + timbre_7_min + timbre_10_min + 
      timbre_10_max + timbre_11_min + timbre_11_max, family = binomial, 
    data = newTrain)
summary(SongsLog4) 










#gradient descent
scale = function(x){
  scalex = (x-mean(x))/(max(x)-min(x))
  return(scalex)
}


nonvars = c("year","songtitle","artistname","songID","artistID","loudness")
SongsTrain = SongsTrain[,!(names(SongsTrain) %in% nonvars)]

x = as.matrix(SongsTrain[,-33])
nscalex = cbind(x0 = 1,x)
newx = apply(x,2,scale)
scalex = cbind(x0 = 1,newx)

y = SongsTrain$Top10
m = length(SongsTrain$timesignature)
theta = as.matrix(rep(1,times=33))
alpha = 0.8

gradient = function(theta,alpha,x,y,m){
h = 1/(1+exp(-x %*% theta))
e = h-y
i = 1
jfunc = c()
cost = -(1/m)*(t(y) %*% log(h) + t(1-y) %*% log(1-h))
dcost = (1/m)*(t(x) %*% e)
jfunc[i] = cost
theta = theta - alpha * dcost
i = i + 1
while(TRUE){
  h = 1/(1+exp(-x %*% theta))
  e = h - y
  cost = -(1/m)*(t(y) %*% log(h) + t(1-y) %*% log(1-h))
  dcost = (1/m)*(t(x) %*% e) 
  jfunc[i] = cost
  if(abs(jfunc[i] - jfunc[i-1])<=1e-15) break
  theta = theta - alpha * dcost
  i = i + 1
}
return(list(theta=theta,cost=cost))
}
theta1 = as.matrix(SongsLog2$coefficients)
h1 = 1/(1+exp(-nscalex %*% theta1))
cost = -(1/m)*(t(y) %*% log(h1) + t(1-y) %*% log(1-h1)) 
#answer:0.3382759

gradient(theta,alpha = 0.1,scalex,y,m)




h = 1/(1+exp(-scalex %*% theta))
e = h-y
i = 1
jfunc = c()
cost = -(1/m)*(t(y) %*% log(h) + t(1-y) %*% log(1-h))
dcost = (1/m)*(t(scalex) %*% e)
jfunc[i] = cost
alpha = 1
theta = theta - alpha * dcost
i = i + 1
while(TRUE){
  h = 1/(1+exp(-scalex %*% theta))
  e = h - y
  cost = -(1/m)*(t(y) %*% log(h) + t(1-y) %*% log(1-h))
  dcost = (1/m)*(t(scalex) %*% e) 
  jfunc[i] = cost
  print(cost)
  if(abs(jfunc[i] - jfunc[i-1])<=1e-15) break
  theta = theta - alpha * dcost
  i = i + 1
}









